src/formatting_stack/core.clj

```diff
@@ -47,30 +48,51 @@
                              :level     :exception}])))))
            (doall)))))
 
-(defn format! [& {:keys [strategies
-                         third-party-indent-specs
-                         formatters
-                         linters
-                         processors
-                         reporter
-                         in-background?]}]
+(speced/defn format! [& {:keys                                             [^vector? strategies
+                                                                            ;; XXX rename `formatters` to `formatter-factories`
+                                                                            ;; and forbid `formatters`.
+                                                                            ;; this will avoid confusion from the legacy API.
+                                                                            ;; same for `linters`, `processors`
+                                                                            ^::protocols.formatter/formatter-factory formatters
+                                                                            ^::protocols.linter/linter-factory linters
+                                                                            ^::protocols.processor/processor-factory processors
+                                                                            third-party-indent-specs
+                                                                            reporter
+                                                                            in-background?]
+                         {formatter-overrides                 :formatters
+                          linter-overrides                    :linters
+                          processor-overrides                 :processors
+                          {formatters-strategies :formatters
+                           linters-strategies    :linters
+                           processors-strategies :processors} :strategies} :overrides}]
```

Comment:
Can we deal with the overrides one level higher e.g. create a fn which applies overrides, then call `format!`? 

That way `format!` does less, and doesn't have to change shape, except for the removal of defaults (which is a good thing imo)